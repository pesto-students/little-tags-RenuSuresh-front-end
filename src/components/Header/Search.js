import React, { useState, useEffect, useRef } from "react";
import { useHistory } from "react-router-dom";
import InputBase from "@material-ui/core/InputBase";
import SearchIcon from "@material-ui/icons/Search";
import { fade, makeStyles } from "@material-ui/core/styles";
import "./Header.css";
import option from "./searchOption.json";

const useStyles = makeStyles((theme) => ({
  search: {
    position: "relative",
    borderRadius: theme.shape.borderRadius,
    backgroundColor: fade(theme.palette.common.white, 0.15),
    "&:hover": {
      backgroundColor: fade(theme.palette.common.white, 0.25),
    },
    marginRight: theme.spacing(2),
    marginLeft: 0,
    height: "2.25rem",
    display: "flex",
    [theme.breakpoints.up("sm")]: {
      marginLeft: theme.spacing(3),
      width: "auto",
      height: "auto",
    },
  },
  searchIcon: {
    padding: theme.spacing(0, 2),
    height: "100%",
    position: "absolute",
    pointerEvents: "none",
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
  },
  inputRoot: {
    color: "inherit",
  },
  inputInput: {
    padding: theme.spacing(1, 1, 1, 0),
    paddingLeft: `calc(1em + ${theme.spacing(4)}px)`,
    transition: theme.transitions.create("width"),
    width: "100%",
    [theme.breakpoints.up("lg")]: {
      width: "40rem",
    },
  },
  sectionDesktop: {
    display: "none",
    [theme.breakpoints.up("md")]: {
      display: "flex",
    },
  },
  sectionMobile: {
    display: "flex",
    [theme.breakpoints.up("md")]: {
      display: "none",
    },
  },
}));
function Search() {
  const classes = useStyles();
  const history = useHistory();
  const [search, setSearch] = useState("");
  const [display, setDisplay] = useState(false);
  const wrapperRef = useRef(null);

  const handleSearch = () => {
    history.push(`/search?title=${search}`);
    setSearch("");
  };

  const inputSearch = ({ target: { value } }) => {
    setSearch(value);

    if (value.length > 1) {
      setDisplay(true);
    } else {
      setDisplay(false);
    }
  };

  const handleClickOutside = (event) => {
    const { current: wrap } = wrapperRef;
    if (wrap && !wrap.contains(event.target)) {
      setDisplay(false);
    }
  };

  const updateSearchDropDown = (value) => {
    setSearch(value);
    setDisplay(false);
  };
  const searchDropDown = (
    <div className="header__searchdropdown">
      {option
        .filter(
          ({ name }) => name.toLowerCase().indexOf(search.toLowerCase()) > -1
        )
        .map((value, i) => {
          if (i > 2) {
            return null;
          }
          return (
            <div
              onClick={() => updateSearchDropDown(value.name)}
              className="option"
              key={i}
              tabIndex="0"
            >
              <span>{value.name}</span>
            </div>
          );
        })}
    </div>
  );
  useEffect(() => {
    window.addEventListener("mousedown", handleClickOutside);
    return () => {
      window.removeEventListener("mousedown", handleClickOutside);
    };
  }, []);

  return (
    <div ref={wrapperRef} className="search">
      <div
        className={classes.search}
        style={{
          background: "#f4f1f1",
          boxShadow: "0px 4px 4px rgb(0 0 0 / 25%)",
        }}
      >
        <InputBase
          placeholder="Searchâ€¦"
          classes={{
            root: classes.inputRoot,
            input: classes.inputInput,
          }}
          inputProps={{ "aria-label": "search" }}
          onChange={inputSearch}
          value={search}
        />
        <span onClick={handleSearch} className="header__search__button">
          <SearchIcon />
        </span>
      </div>
      {display && searchDropDown}
    </div>
  );
}

export default Search;
