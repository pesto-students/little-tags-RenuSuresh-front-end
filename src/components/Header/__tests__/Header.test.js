import { fireEvent, render, cleanup } from "@testing-library/react";
import React from "react";
import Enzyme from "enzyme";
import renderer from "react-test-renderer";
import Search from "../Search";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
Enzyme.configure({ adapter: new Adapter() });

const { mount } = Enzyme;

afterEach(cleanup);

describe("<Search />", () => {
  it("should render snapshot", () => {
    const tree = renderer.create(<Search />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Should have one input field", () => {
    const wrapper = mount(<Search />);
    expect(wrapper.find("input")).toHaveLength(1);
  });

  it("should have empty input filed", () => {
    const wrapper = mount(<Search />);
    expect(wrapper.find("input").props().value).toEqual("");
  });
  it("should not display suggestion on click/focus in search input", () => {
    const tree = mount(<Search />);
    tree.find("input").simulate("click");
    expect(tree).toMatchSnapshot();
  });
  it("should update input text", () => {
    const tree = mount(<Search />);
    tree.find("input").simulate("change", { target: { value: "jac" } });
    expect(tree.find("input").props().value).toEqual("jac");
  });
});
