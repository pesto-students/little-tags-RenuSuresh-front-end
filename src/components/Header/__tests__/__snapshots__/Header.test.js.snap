// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> renders when there are no items 1`] = `
<div
  className="search"
>
  <div
    className="makeStyles-search-1"
    style={
      Object {
        "background": "#f4f1f1",
        "boxShadow": "0px 4px 4px rgb(0 0 0 / 25%)",
      }
    }
  >
    <div
      className="MuiInputBase-root makeStyles-inputRoot-3"
      onClick={[Function]}
    >
      <input
        aria-label="search"
        className="MuiInputBase-input makeStyles-inputInput-4"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Search…"
        type="text"
        value=""
      />
    </div>
    <span
      className="header__search__button"
      onClick={[Function]}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
    </span>
  </div>
</div>
`;

exports[`<Search /> should not display suggestion on click/focus in search input 1`] = `ReactWrapper {}`;

exports[`<Search /> should render snapshot 1`] = `
<div
  className="search"
>
  <div
    className="makeStyles-search-1"
    style={
      Object {
        "background": "#f4f1f1",
        "boxShadow": "0px 4px 4px rgb(0 0 0 / 25%)",
      }
    }
  >
    <div
      className="MuiInputBase-root makeStyles-inputRoot-3"
      onClick={[Function]}
    >
      <input
        aria-label="search"
        className="MuiInputBase-input makeStyles-inputInput-4"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Search…"
        type="text"
        value=""
      />
    </div>
    <span
      className="header__search__button"
      onClick={[Function]}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
    </span>
  </div>
</div>
`;

exports[`<Search /> should update input text on useState 1`] = `
<div
  className="search"
>
  <div
    className="makeStyles-search-1"
    style={
      Object {
        "background": "#f4f1f1",
        "boxShadow": "0px 4px 4px rgb(0 0 0 / 25%)",
      }
    }
  >
    <div
      className="MuiInputBase-root makeStyles-inputRoot-3"
      onClick={[Function]}
    >
      <input
        aria-label="search"
        className="MuiInputBase-input makeStyles-inputInput-4"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Search…"
        type="text"
        value=""
      />
    </div>
    <span
      className="header__search__button"
      onClick={[Function]}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
    </span>
  </div>
</div>
`;
